subsys/storage/storage_area/storage_area_store.c

```diff
@@ -0,0 +1,1097 @@
+}
+
+static ALWAYS_INLINE int
+store_take_semaphore(const struct storage_area_store *store)
+{
+#ifdef CONFIG_STORAGE_AREA_STORE_SEMAPHORE
+	return k_sem_take(&store->data->semaphore, K_FOREVER);
+#else
+	ARG_UNUSED(store);
+	return 0;
+#endif /* CONFIG_STORAGE_AREA_STORE_SEMAPHORE */
+}
+
+static ALWAYS_INLINE void
+store_give_semaphore(const struct storage_area_store *store)
+{
+#ifdef CONFIG_STORAGE_AREA_STORE_SEMAPHORE
+	(void)k_sem_give(&store->data->semaphore);
+#endif /* CONFIG_STORAGE_AREA_STORE_SEMAPHORE */
+}
+
+static bool store_record_valid(const struct storage_area_record *record)
+{
+	const struct storage_area *area = record->store->area;
+	const size_t crc_skip = record->store->crc_skip;
+	const size_t recpos = record->sector * record->store->sector_size +
+			      record->loc + SAS_HDRSIZE + crc_skip;
+	uint32_t crc = SAS_CRCINIT;
+	uint8_t buf[SAS_MAX(SAS_MINBUFSIZE, area->write_size)];
+	struct storage_area_iovec rd = {
+		.data = &buf,
+	};
+	size_t rdlen = record->size - crc_skip;
+	sa_off_t rdoff = recpos;
+
+	while (rdlen != 0U) {
+		rd.len = SAS_MIN(sizeof(buf), rdlen);
+		if (storage_area_readv(area, rdoff, &rd, 1U) != 0) {
+			goto end;
+		}
+
+		crc = crc32_ieee_update(crc, buf, rd.len);
+		rdlen -= rd.len;
+		rdoff += rd.len;
+	}
+
+	rd.len = SAS_CRCSIZE;
+	if (storage_area_readv(area, rdoff, &rd, 1U) != 0) {
+		goto end;
+	}
+
+	if (crc != sys_get_le32(buf)) {
+		goto end;
+	}
+
+	return true;
+end:
+	LOG_DBG("record at [%d-%d] has bad crc", record->sector, record->loc);
+	return false;
+}
+
+static int store_record_next_in_sector(struct storage_area_record *record,
+				       bool wrapcheck, bool recover)
+{
+	const struct storage_area_store *store = record->store;
+	const struct storage_area_store_data *data = store->data;
+	const struct storage_area *area = store->area;
+	const size_t secpos = record->sector * store->sector_size;
+	bool crc_ok = true;
+	int rc = -ENOENT;
+
+	if ((record->loc == 0U) && (store->sector_cookie != NULL) &&
+	    (store->sector_cookie_size != 0U)) {
+		const size_t cksz = store->sector_cookie_size;
+
+		record->loc = SAS_ALIGNUP(cksz, area->write_size);
+	}
+
+	while (true) {
+		uint8_t buf[SAS_HDRSIZE];
+		struct storage_area_iovec rd = {
+			.data = &buf,
+			.len = sizeof(buf),
+		};
+		size_t rdpos = record->loc;
+		sa_off_t rdoff;
+
+		if (record->size != 0U) {
+			rdpos += (SAS_HDRSIZE + record->size + SAS_CRCSIZE);
+			rdpos = SAS_ALIGNUP(rdpos, area->write_size);
+		}
+
+		if (((data->sector == record->sector) && (data->loc <= rdpos)) ||
+		    (rdpos == store->sector_size)) {
+			record->loc = rdpos;
+			record->size = 0U;
+			rc = -ENOENT;
+			break;
+		}
+
+		rdoff = secpos + rdpos;
+		rc = storage_area_readv(area, rdoff, &rd, 1U);
+		if (rc != 0) {
+			break;
+		}
+
+		size_t rsize = (size_t)sys_get_le16(&buf[2]);
+		size_t avail =
+			store->sector_size - rdpos - SAS_CRCSIZE - SAS_HDRSIZE;
+		bool size_ok = ((rsize > 0U) && (rsize < avail));
```

Comment:
There is no hidden logic, the system does not care about the erase-value.

I could add a comment that the MAGIC was chosen as not to be found on a system with erase-value 0x00 or 0xff,

Reply:
fixed

Reply:
Oh, I thought this works as planned, and you just have not noticed:
0. crc_ok = true, record->size = don't care, wrapchk = false
1. at line 150 rdoff = some sector offset + rdpos (don't care)
2. at 151 header of record gets read
3. at 156, without confirming that is valid record header, size is read with  higher limit of  0xffff

Now at line 157 `avail` is calculated, which I guess is limited at whatever `size_t` can handle - `SAS_CRCSIZE + SAS_HDRSIZE`.

Now assume that avail is < 0xffff, at line 159 `size_ok` will be `false` on both types of devices, erasing to 0xff and 0x00, as `((rsize > 0U) && (rsize < avail))` for `size == 0x0000' or `rsize == 0xffff` - helping with detecting that there is no record, kinda.

But as soon as `sector_size` gets bigger than 0xffff this becomes true for any `rsize` != 0, as `rsize` is limited at 0xffff.
Above and any junk that fits in the range will make the `size_ok == true`.

Now in line 165, wrapcnt of record gets overwritten.
Condition in line 169 is at this point false, because, from point 0, `crc_ok == true` and from above `size_ok == true`,
which means that `store_record_valid` is not run on the record.

If there is now junk that is equivalent of MAGIC in position of magic, from above `crc_ok == true`, `size_ok == true`, wrap counter is the same as set in line 165, the function returns that it found record, even though it has no idea what it has found.

Reply:
Hi @de-nordic, sorry for the late reply.

The aim of this routine is to find the next "valid" entry. A valid entry will have the following properties:
1. The header magic and wrapcnt are correct,
2. The size is valid (> 0 and does not cross a sector boundary),

When everything is OK also the crc32 will be correct, but for speed reasons it is desired to postpone the crc32 verification to the user. In "normal" situations the crc32 is not checked. However if the walking needs to recover from bad data (in magic, wrapcnt, size) no matter how they have appeared, the crc32 check is enabled as an extra check. 

I am reworking the logic a little.